Loader for new D16 data

MathieuTillet opened the pull request:
**Description of work.**
This PR makes it possible for Mantid to manage the files D16 will output once its detector is changed next cycle. The format does not fundamentally change, but scans are now grouped into a multidimensional file using the TOF dimension for the scanned parameter.

Reduction algorithms usually used by D16 are also updated to reflect this change.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**
-  Load a test file ( such as `UnitTest/ILL/D16/066321.nxs` which is mostly white noise except for one corner) and check it is well formed
- Open the instrument viewer and check the view looks fine (especially in 3D with the proportions correct)

<!-- Instructions for testing. -->

Closes #34163  <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

MathieuTillet committed:
[D16] Loader prototype for new geometry

Loads the data currently defined in nexus files, albeit quite uglyly

MathieuTillet committed:
Manage dimension finding better in LoadILLSANS

Determinign which dimension of the data correspond to which physical
reality is a mess that depends on instruments and versions of the
instruments, so this should make things clearer to read and maintain, if
only a little

MathieuTillet committed:
Load D16 omega binning

MathieuTillet committed:
Add correct unit for D16 scans

MathieuTillet committed:
New version for D16 parameter scan

MathieuTillet committed:
Add a parameter file for D16B

MathieuTillet committed:
Fix docstring to follow PEP

So Pycharm will stop complaining about that

MathieuTillet committed:
Remove unused wavelength parameter

No longer useful since the files are now well built

MathieuTillet committed:
Refactoring of PyExec

MathieuTillet committed:
Cleaner vector handling

MathieuTillet committed:
Fix normalisation parameter name

MathieuTillet committed:
Fix distance default for new d16

MathieuTillet committed:
[D16] Fix pixel sizes

MathieuTillet committed:
Load D16B monitor values

MathieuTillet committed:
Correctly determine number of monitors for D16

Which is 1, not 2.

MathieuTillet committed:
Fix type issues for new parameters

MathieuTillet committed:
Update test for new version of the algorihtm

The old one is now longer used - it never really was - so it is just
thrown to the bin and the old test is no longer relevant

MathieuTillet committed:
Add D16 new definition file

MathieuTillet committed:
Add test data

MathieuTillet committed:
Update documentation for LoadILLSANS and SANSILLParameterScan

MathieuTillet committed:
Add test for LoadILLSANS

MathieuTillet committed:
Add release notes

MathieuTillet committed:
Put some parameters const to please cppcheck

And the reviewer

darominski reviewed (changes requested):
The new D16B data is loaded and looks sensibly, also the instrument in the instrument viewer looks fine, without overlaps and other spurious effects. This is fine provided that the new instrument is indeed a 'banana' instrument and it is intended that the first pixel is in the bottom-right corner when looked at in Cylindrical Y coordinates. I have some comments about the code and they are dispersed throughout, and below you can find a couple comments and requests about the functionality:

- Could you add a confidence test to the LoadILLTest of the D16B data?
- It is not possible to load empty D16B instrument, as D16B was not added to the list of instruments for our facility, could you add it?

MathieuTillet commented on Framework/DataHandling/src/LoadILLSANS.cpp:
```diff
@@ -560,40 +648,52 @@ size_t LoadILLSANS::loadDataFromMonitors(NeXus::NXEntry &firstEntry, size_t firs
-      m_localWorkspace->setCountVariances(index, histoVariances);
+  if (m_instrumentName == "D16B") {
+    PARALLEL_FOR_IF(Kernel::threadSafe(*m_localWorkspace))
+    for (int i = 0; i < numberOfTubes; i++) {
```
Good point, this entire part is starting to be a bit hard to read so that could help.

MathieuTillet committed:
Clarify loader code with comments and better names

MathieuTillet committed:
Catch runtime error when no distance provided

MathieuTillet commented on Framework/DataHandling/src/LoadILLSANS.cpp:
```diff
@@ -127,7 +124,12 @@ void LoadILLSANS::exec() {
+    double distance;
+    try {
+      distance = firstEntry.getFloat(instrumentPath + "/Det/value") / 1000; // mm to metre
+    } catch (...) {
```
It can, and now is.

MathieuTillet committed:
Change default for pixel range basd on new geometry

MathieuTillet committed:
Get normalisation options from SANSILLReduction

Copying rather than rewriting the option. For that, changing the
algorithm to a more precise DataProcessorAlgorithm is necessary

MathieuTillet committed:
Cleanup useless code and old todos

MathieuTillet commented on docs/source/algorithms/LoadILLSANS-v1.rst:
```diff
@@ -9,13 +9,22 @@
 
-This loads the nexus files produced by the SANS instruments D11, D22, D33 at the ILL.
-For D33, it supports both monochromatic and TOF modes.
+This loads the nexus files produced by the SANS instruments D11, D22, D33 and D16 at the ILL.
```
That is covered by the mention later that it also supports newer version of those instruments, since D22B also exists and the list will start to be a bit long.

MathieuTillet committed:
Test that D16B data triggers the correct loader

MathieuTillet committed:
Add D16B to the list of suppoorted instruments

darominski committed:
Add extension to SampleRun

FileProperty seems to expect the exact file name, including the extension to be able to find the file.

darominski commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLParameterScan.py:
```diff
@@ -137,101 +130,122 @@ def PyExec(self):
-        # Omega scans are loaded as histogram data for compatibility reasons, but we need point data from now on.
-        # Converting from histogram to point data is straight forward and no meaningful information can be lost.
-        ConvertToPointData(InputWorkspace=load_ws_name + "_grouped", OutputWorkspace=load_ws_name + "_grouped")
+        Load(Filename=self.sample, OutputWorkspace=load_ws_name, startProgress=0, endProgress=0.7)
```
All right, let's leave it then.

darominski reviewed (dismissed):
Thanks for all the changes and addressing the comments. As agreed, I committed the suggested change in the doc test. This looks good to go now, as soon as tests are passing.

MathieuTillet committed:
Force double in std::accumulate

darominski reviewed (approved):
Re-approved. Thank you for fixing the `std::accumulate` issue of the Windows' compiler.
